### PR TITLE
CHE-220: Improving 'removeContainer' API. Refactoring

### PR DESCRIPTION
**Pull request by ibuziuk**
### What does this PR do?
Enhance OpenShift resource deletion for workspaces (scaling deployment down before removing resources). Refactoring

### What issues does this PR fix or reference?
https://issues.jboss.org/browse/CHE-220

### CODE REVIEW

**Comment by ibuziuk** [comment]
cc @mlabuda

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftConnector.java [hunk 1]
```diff
@@ -158,8 +155,9 @@
 @Singleton
 public class OpenShiftConnector extends DockerConnector {
     private static final Logger LOG                                      = LoggerFactory.getLogger(OpenShiftConnector.class);
-    public static final String CHE_OPENSHIFT_RESOURCES_PREFIX           = "che-ws-";
-    public static final String OPENSHIFT_DEPLOYMENT_LABEL               = "deployment";
+    public static final String CHE_OPENSHIFT_RESOURCES_PREFIX            = "che-ws-";
+    public static final String OPENSHIFT_DEPLOYMENT_LABEL                = "deployment";
+    public static final int OPENSHIFT_WAIT_POD_TIMEOUT                   = 240;
```
Comment (sunix) [inline plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftConnector.java:160]: are these ones real changes or just formatting ?

Comment (ibuziuk) [reply plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftConnector.java:160]: real change & 2 deliberate shifts of '=' sign

**Review by mlabuda — CHANGES_REQUESTED** [review]
One comment, otherwise LGTM.

**Review by sunix — APPROVED** [review]
LGTM

**Comment by ibuziuk** [comment]
@mlabuda there is no comment ;-)

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftContainerRemover.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import io.fabric8.openshift.api.model.Route;
+import io.fabric8.openshift.client.DefaultOpenShiftClient;
+import io.fabric8.openshift.client.OpenShiftClient;
+
+@Singleton
+public class OpenShiftContainerRemover {
+    private static final Logger LOG = LoggerFactory.getLogger(OpenShiftContainerRemover.class);
+
+    public void removeContainer(final String deploymentName, final String namespace) throws IOException {
+        scaleDownDeployment(deploymentName, namespace);
```
Comment (mlabuda) [inline plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftContainerRemover.java]: did you check that waiting in scaleDownDeployment works? If you remember, there is that bug with k8 client in deployment config waitings (https://github.com/fabric8io/kubernetes-client/issues/699). If it is same for deployments, then it waits for fixed time and then continue...

**Review by mlabuda — CHANGES_REQUESTED** [review]
Ah, my comment somehow disappeared between posting a comment and submitting review. See comment now.

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftContainerRemover.java (hunk 2 above)
Comment (ibuziuk) [reply plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftContainerRemover.java]: @mlabuda it works for me just fine

Comment (mlabuda) [reply plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftContainerRemover.java]: If it works, +1 ;)

**Review by mlabuda — APPROVED** [review]

**Review by l0rd — COMMENTED** [review]
Great job @ibuziuk but I would rename `OpenShiftContainerRemover` class and `removeContainer()` method because we are not removing a container but a deployment. Maybe `OpenShiftDeploymentCleaner` and `cleanDeploymentResources()` are more indicated

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftContainerRemover.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.dsl.FilterWatchListDeletable;
+import io.fabric8.kubernetes.client.dsl.ScalableResource;
+import io.fabric8.openshift.api.model.Route;
+import io.fabric8.openshift.client.DefaultOpenShiftClient;
+import io.fabric8.openshift.client.OpenShiftClient;
+
+@Singleton
+public class OpenShiftContainerRemover {
```
Comment (amisevsk) [inline plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftContainerRemover.java]: WDYT about making this class a static utility class? That way we don't have to inject it in OpenShiftConnector, and can instead refer to static methods. The OpenShiftConnector constructor is already getting quite unwieldy.

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftContainerRemover.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            }
+
+            if (service != null) {
+                LOG.info("Removing OpenShift Service {}", service.getMetadata().getName());
+                openShiftClient.resource(service).delete();
+            }
+        }
+    }
+    
+    private void waitUntilWorspacePodIsDeleted(final String deploymentName, final String namespace) throws OpenShiftException {
```
Comment (amisevsk) [inline plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftContainerRemover.java]: Typo: waitUntilWor**k**spacePodIsDeleted

File: plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftRouteCreator.java [hunk 5]
```diff
@@ -19,17 +21,18 @@
...
     private static final String TLS_TERMINATION_EDGE = "edge";
     private static final String REDIRECT_INSECURE_EDGE_TERMINATION_POLICY = "Redirect";
 
-    public static void createRoute (final String namespace,
-                                    final String workspaceName,
-                                    final String openShiftNamespaceExternalAddress,
-                                    final String serverRef,
-                                    final String serviceName,
-                                    final boolean enableTls) {
+    public void createRoute (final String namespace,
```
Comment (amisevsk) [inline plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftRouteCreator.java:30]: What do we gain by making this class non-static?

**Review by amisevsk — CHANGES_REQUESTED** [review]
One typo that should be fixed, looks good to me apart from that. I've added a few comments for clarification as well.

**Comment by amisevsk** [comment]
+1 to the rename suggested by @l0rd

**Review by amisevsk — APPROVED** [review]

**Comment by ibuziuk** [comment]
@amisevsk @l0rd rename & typo are fixed
in terms of static vs. singleton components (in our context it can be also rephrased to procedure programming vs OOP)

Basically, the rule-of-thumb is that static methods should be used for pure utilities like `KubernetesResourceUtil.java` and code reuse should be achieved via composition in OOP (via DI in our case).

However, I do agree that with the increasing number of components the constructor is becoming polluted with the args.  In Spring, for example, there is no such problem with polluted constructors - 
`@Autowired` and `@Value` annotations are used directly on class fields for injecting components / env vars. Maybe, it is possible to reimplement DI in OpenShiftConnector and inject properties without using constructor ? 

In general, I would  prefer to use `@Singleton` annotation for stateless components as much as possible

**Comment by amisevsk** [comment]
@ibuziuk Alright, makes sense. :+1:

**Review by l0rd — APPROVED** [review]

**Comment by l0rd** [comment]
If @sunix is ok we can merge

**Comment by ibuziuk** [comment]
cherry-picking to `openshift-connector-rebased` https://github.com/sunix/che/pull/1

**Comment by ibuziuk** [comment]
cherry-picking to origin's `openshift-connector-rebased` https://github.com/eclipse/che/pull/5191

**Pull request merged**